design/events.md

```diff
@@ -0,0 +1,439 @@
+new system, as explained below.
+
+And finally, the present document assumes that Nessie's new storage model will be adopted before the work on this
+feature starts, and thus takes the new model into account whenever appropriate.
+
+# 2. Design principles
+
+## 2.1. Events
+
+An event, in the scope of this document, is defined as _the act of modifying the catalog state_.
+
+First corollary of this definition: a read of the catalog state is not an event.
+
+Also, an event should only be triggered if the catalog was effectively modified. E.g., if a commit operation could not 
+be carried out due to a concurrent modification of the branch, no event should be raised, since the operation did not 
+modify the catalog state (but the concurrent modification, if succeeded, triggered such an event).
+
+A complete list of supported event types is given below, but it's important to first discuss the different event 
+categories that should be covered by the new system:
+
+### 2.1.1. "Git-like" events
+
+Since Nessie is a Git-like transactional catalog at its core, events in this category are Nessie's primary events.
+
+Git-like events can be further divided into 2 sub-categories:
+
+1. **Reference events**: a reference event is the act of manipulating a named reference; in other words, it's the act of
+   moving or creating pointers to existing hashes. These are:
+   1. Branch events: when a branch is created, reassigned to a new hash, or deleted;
+   2. Tag events: when a tag is created, reassigned to a new hash, or deleted;
+2. **"Committing" events**: the act of adding new hashes to the catalog. There are three concrete committing events:
+   1. Commit events: probably the most central event type, the commit event is triggered by an API call to a commit
+      endpoint.
+   2. Merge events: a merge event is triggered by an API call to a merge endpoint.
+   3. Transplant events: a transplant event is triggered by an API call to a transplant endpoint.
+
+Notes:
+
+* Git-like events are rather "low-level" events. Most subscribers of the new system would probably not subscribe
+  directly to these events, but rather to more high-level events (see below); however, Git-like events could prove
+  useful for audit use cases ("who did what and when").
+* Git-like events have a close relationship to API calls and are almost always triggered by such calls. Only a few 
+  Git-like events are exceptionally triggered outside an API call, e.g. when the default branch is initialized.
+* All committing events can reference multiple operations (PUTs and DELETEs), and theoretically, multiple tables or 
+  views. Therefore, a committing event should be accompanied by zero, one or multiple content-related events, one for 
+  each table or view affected; see below. (Zero content events are possible, e.g. with empty merges.)
+* The new storage model opens up the notion of "references" to be a generic pointer to any kind of stored object. 
```

Comment:
Is the full reference name useful for end users? I'm not sure. Since it's not a notion available in the old storage model, I wonder if we should include it.
As for omitting the distinction between reference types for events, that's a good idea! Because it makes the API more extensible and resilient to new reference types. In this case we'd need to change the events to `REFERENCE_(CREATED|UPDATED|DELETED)` and add a `ReferenceType` field. Would that be OK for you?

Reply:
The old storage model will eventually disappear.
End users won't interact directly with events, there'll always be something "in between".

The advantage of having the "full ref names" would be that other systems could react to changes to custom references.
Think: references for pull-requests or so.

Reply:
OK. In this case, I will amend the events type list and mention that reference names should come in full form. (Maybe short and full form together?)